src/expand.rs

```diff
@@ -247,247 +268,86 @@ fn transform_sig(
+                if ident == "self" {
+                    self_span = Some(ident.span());
+                    let prefixed = format_ident!("{}{}", self_prefix, ident);
+                    quote!(let #mutability #prefixed = #ident;)
+                } else {
+                    quote!(let #mutability #ident = #ident;)
                 }
+            } else {
+                let pat = &arg.pat;
+                let ident = positional_arg(i, pat.span());
+                quote!(let #pat = #ident;)
             }
         }
-        _ => {}
-    }
-
-    if let Context::Trait { name, generics, .. } = context {
-        if has_self {
-            let (_, generics, _) = generics.split_for_impl();
-            let mut self_param: TypeParam = parse_quote!(AsyncTrait: ?Sized + #name #generics);
-            if !is_local {
-                self_param.bounds.extend(self_bound);
-            }
-            let count = standalone
-                .generics
-                .params
-                .iter()
-                .take_while(|param| {
-                    if let GenericParam::Const(_) = param {
-                        false
-                    } else {
-                        true
-                    }
-                })
-                .count();
-            standalone
-                .generics
-                .params
-                .insert(count, GenericParam::Type(self_param));
-            types.push(Ident::new("Self", Span::call_site()));
-        }
-    }
+    }).collect::<Vec<_>>();
 
-    if let Some(where_clause) = &mut standalone.generics.where_clause {
-        // Work around an input bound like `where Self::Output: Send` expanding
-        // to `where <AsyncTrait>::Output: Send` which is illegal syntax because
-        // `where<T>` is reserved for future use... :(
-        where_clause.predicates.insert(0, parse_quote!((): Sized));
+    if let Some(span) = self_span {
+        let mut replace_self = ReplaceSelf(self_prefix, span);
+        replace_self.visit_block_mut(block);
     }
 
-    let mut replace = match context {
-        Context::Trait { .. } => ReplaceReceiver::with(parse_quote!(AsyncTrait)),
-        Context::Impl {
-            receiver, as_trait, ..
-        } => ReplaceReceiver::with_as_trait(receiver.clone(), as_trait.clone()),
+    let stmts = &block.stmts;
+    let ret_ty = match &sig.output {
+        ReturnType::Default => quote_spanned!(block.span()=>()),
+        ReturnType::Type(_, ret) => quote!(#ret),
     };
-    replace.visit_signature_mut(&mut standalone);
-    replace.visit_block_mut(block);
 
-    let mut generics = types;
-    let consts = standalone
-        .generics
-        .const_params()
-        .map(|param| param.ident.clone());
-    generics.extend(consts);
+    let box_pin = quote_spanned!(ret_ty.span()=>
+        Box::pin(async move {
+            let __ret: #ret_ty = {
+                #(#decls)*
+                #(#stmts)*
```

Comment:
In order for https://github.com/tokio-rs/tracing/pull/1228 to work well, could you consider adding `let __async_trait: ();` between the `#(#decls)*` and `#(#stmts)*` lines? That should do the job. Thanks!

Reply:
I think this is a poor solution to the problem at hand, and that emitting warnings is, in fact, the desired effect. Nevertheless, I've pushed a commit that implements this.

Reply:
I do agree that it's a hack, and a nasty one at that. On the other hand, I do not believe that users will understand that exhibit A will trigger a warning and not exhibit B. That doesn't appear very coherent (at least with what I think is the behavior of `instrument` in non-async-trait functions) either.

Exhibit A:
```rust
#[instrument(fields(test=%v+5))]
async fn call(&mut self, v: usize) {}
```

Exhibit B:
```rust
#[instrument(fields(test=%_v+5))]
async fn call(&mut self, _v: usize) {}
```

Reply:
I disagree that users won't undestand. Without the attribute, `A` triggers a warning and `B` does not because `v` is unused in the body of `A`. Even when the attribute is applied, `v` continues to be unused in `A`, and so it is my opinion that `A` should continue to result in a warning.

I don't mean to imply that `instrument` should inconsistently emit warnings based on whether `async-trait` is used, but instead, that it should consistently emit warnings as would be emitted without the attribute.